### PR TITLE
Adding PageActionDescriptorProvider

### LINKED ISSUE #5353: Routing for View Pages
## Routing

By default, the route for each page is mapped at it's application relative path (with the cshtml extension). The file name `Index` is optional and has the same behavior as routing to an action named `Index` in conventional routing i.e.
/Home/Index and /Home are equivalent. The optional value of the`@page` directive can be used to append a suffix to this route. Consider,

| Page | Directive | Routes |
| --- | --- | --- |
| /About.cshtml | `@page` | http://myappdomain.net/About |
| /Contact.cshtml | `@page "location/{zip}"` | http://myappdomain.net/Contact/location/98052 |
| /Catalog/Index.cshtml | `@page` | http://myappdomain.net/Catalog http://myappdomain.net/Catalog/Index |
| /Catalog/Books.cshtml | `@page "{id?}"` | http://myappdomain.net/Catalog/Books/ http://myappdomain.net/Catalog/Books/12 |
## Friendly urls

Associated with each page is an application model type similar to `ControllerModel`. We'd build on top of this to provide simple runtime configuration of routes. For instance, to configure an additional route to the `/Catalog/Books.cshtml`, here's what the Startup code might look like:

```
services
    .AddMvc()
    .AddRazorPages(options =>
    {
        options.MapPageRoute("Books/{id?}", "/Catalog/Books.cshtml");
    });
```

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #5353

### CODE REVIEW

**Comment by pranavkm** [comment]
Resending because https://github.com/aspnet/Mvc/pull/5450 was messed up. cc @rynowak \ @javiercn

**Comment by pranavkm** [comment]
Bump, I'd like to get this in soon.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageModels.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc.Filters;
+
+namespace Microsoft.AspNetCore.Mvc.ApplicationModels
+{
+    /// <summary>
+    /// Application model component for RazorPages.
+    /// </summary>
+    public class PageModel
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageModels.cs]: Filename here is wrong, `PageModels.cs`

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageActionDescriptorProviderTest.cs [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+            var provider = new PageActionDescriptorProvider(
+                razorProject.Object,
+                GetAccessor<MvcOptions>(),
+                GetAccessor<RazorPagesOptions>());
+            var context = new ActionDescriptorProviderContext();
+
+            // Act and Assert
+            var ex = Assert.Throws<InvalidOperationException>(() => provider.OnProvidersExecuting(context));
+            Assert.Equal(
+                "The @page directive for the Razor page at /Test.cshtml cannot override the relative path prefix.",
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageActionDescriptorProviderTest.cs]: This is a really jargon-ey error. I'd suggest something like
> The route for a page cannot start with `/` or `~/`. Pages do not support overriding the file path of the page.

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/project.json [hunk 3]
```diff
@@ -1,14 +1,19 @@
 ﻿{
   "buildOptions": {
-    "warningsAsErrors": true
+    "warningsAsErrors": true,
+    "compile": {
+      "include": "../Microsoft.AspNetCore.Mvc.TestCommon/TestFile*.cs"
+    }
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/project.json]: What are these?

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageActionDescriptorProviderTest.cs (hunk 2 above)
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageActionDescriptorProviderTest.cs]: Sure

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/project.json (hunk 3 above)
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/project.json]: I think I had some odd issues referencing TestCommon before we had rebased. I'll see if it's fixed now.

**Comment by javiercn** [comment]
Looks good to me, but get @rynowak sign-off

**Comment by rynowak** [comment]
I did 😆

**Pull request closed without merging**